File: matlab/+rotation_converter/Rotation.m
classdef Rotation
%ROTATION Immutable rotation representation with multiple output formats.
%
%   Stores a rotation as a unit quaternion [w, x, y, z] and provides
%   conversion to/from all supported representations.
%
%   Factory methods (static):
%     rot = rotation_converter.Rotation.identity()
%     rot = rotation_converter.Rotation.from_quaternion(q)
%     rot = rotation_converter.Rotation.from_rotation_matrix(R)
%     rot = rotation_converter.Rotation.from_euler(a, b, c, convention)
%     rot = rotation_converter.Rotation.from_axis_angle(axis, angle)
%     rot = rotation_converter.Rotation.from_rodrigues(r)
%
%   Output methods:
%     q = rot.as_quaternion()
%     R = rot.as_rotation_matrix()
%     [a, b, c] = rot.as_euler(convention)
%     [axis, angle] = rot.as_axis_angle()
%     r = rot.as_rodrigues()
%
%   Composition:
%     rot3 = rot1.compose(rot2)
%     rot_inv = rot.inverse()
%
%   Compatible with MATLAB R2014b+ and Octave 5.0+.

    properties (Access = private)
        q_  % internal unit quaternion [w, x, y, z]
    end

    methods
        function obj = Rotation(q)
            %ROTATION Construct from raw unit quaternion (prefer factory methods).
            q = q(:)';
            rotation_converter.internal.require(numel(q) == 4, ...
                'quaternion must have 4 elements');
            obj.q_ = rotation_converter.normalize_quaternion(q);
        end

        % --- Output conversions ---

        function q = as_quaternion(obj)
            %AS_QUATERNION Return [w, x, y, z] unit quaternion (copy).
            q = obj.q_;
        end

        function R = as_rotation_matrix(obj)
            %AS_ROTATION_MATRIX Return 3x3 SO(3) rotation matrix.
            R = rotation_converter.quaternion_to_rotation_matrix(obj.q_);
        end

        function [a, b, c] = as_euler(obj, convention)
            %AS_EULER Return Euler angles for the given convention.
            [a, b, c] = rotation_converter.quaternion_to_euler(obj.q_, convention);
        end

        function [axis, angle] = as_axis_angle(obj)
            %AS_AXIS_ANGLE Return axis (unit vector) and angle (radians).
            [axis, angle] = rotation_converter.quaternion_to_axis_angle(obj.q_);
        end

        function r = as_rodrigues(obj)
            %AS_RODRIGUES Return 3-element Rodrigues vector (axis * angle).
            r = rotation_converter.quaternion_to_rodrigues(obj.q_);
        end

        % --- Composition ---

        function result = compose(obj, other)
            %COMPOSE Return the composition of this rotation with another.
            q_new = rotation_converter.quaternion_multiply(obj.q_, other.q_);
            result = rotation_converter.Rotation(q_new);
        end

        function result = inverse(obj)
            %INVERSE Return the inverse rotation.
            result = rotation_converter.Rotation( ...
                rotation_converter.quaternion_conjugate(obj.q_));
        end

        % --- Display ---

        function disp(obj)
            fprintf('Rotation(q=[%.6f, %.6f, %.6f, %.6f])\n', ...
                    obj.q_(1), obj.q_(2), obj.q_(3), obj.q_(4));
        end

        function result = eq(obj, other)
            %EQ Test equality (up to sign ambiguity).
            if ~isa(other, 'rotation_converter.Rotation')
                result = false;
                return;
            end
            d = abs(dot(obj.q_, other.q_));
            result = abs(d - 1.0) < 1e-10;
        end
    end

    methods (Static)
        function obj = identity()
            %IDENTITY Create identity rotation.
            obj = rotation_converter.Rotation([1, 0, 0, 0]);
        end

        function obj = from_quaternion(q)
            %FROM_QUATERNION Create Rotation from quaternion [w, x, y, z].
            obj = rotation_converter.Rotation(q);
        end

        function obj = from_rotation_matrix(R)
            %FROM_ROTATION_MATRIX Create Rotation from 3x3 SO(3) matrix.
            q = rotation_converter.rotation_matrix_to_quaternion(R);
            obj = rotation_converter.Rotation(q);
        end

        function obj = from_euler(a, b, c, convention)
            %FROM_EULER Create Rotation from Euler angles.
            q = rotation_converter.euler_to_quaternion(a, b, c, convention);
            obj = rotation_converter.Rotation(q);
        end

        function obj = from_axis_angle(axis, angle)
            %FROM_AXIS_ANGLE Create Rotation from axis-angle.
            q = rotation_converter.axis_angle_to_quaternion(axis, angle);
            obj = rotation_converter.Rotation(q);
        end

        function obj = from_rodrigues(r)
            %FROM_RODRIGUES Create Rotation from Rodrigues vector.
            q = rotation_converter.rodrigues_to_quaternion(r);
            obj = rotation_converter.Rotation(q);
        end
    end
end

File: src/shared/python/rotation_transforms/rotation.py
"""Shared rotation representation conversions."""

from __future__ import annotations

from typing import Any

import numpy as np
from scipy.spatial.transform import Rotation as SciPyRotation

from contracts import require, require_finite, require_unit_vector


def _as_finite_array(value: Any, shape: tuple[int, ...], name: str) -> np.ndarray:
    array = np.asarray(value, dtype=float)
    require(array.shape == shape, f"{name} must have shape {shape}", array.shape)
    require_finite(array, name)
    return array


class Rotation:
    """Immutable rotation stored internally as a unit quaternion.

    Public quaternions use the repository convention ``[w, x, y, z]``.
    SciPy uses ``[x, y, z, w]`` internally, so all crossings are normalized
    through small adapters here.
    """

    __slots__ = ("_rotation",)

    def __init__(self, q: Any) -> None:
        q_array = _as_finite_array(q, (4,), "quaternion")
        norm = np.linalg.norm(q_array)
        require(bool(norm > 1e-12), "quaternion must be non-zero", norm)
        q_array = q_array / norm
        if q_array[0] < 0:
            q_array = -q_array
        self._rotation = SciPyRotation.from_quat(
            [q_array[1], q_array[2], q_array[3], q_array[0]]
        )

    @classmethod
    def identity(cls) -> Rotation:
        """Create the identity rotation."""
        return cls([1.0, 0.0, 0.0, 0.0])

    @classmethod
    def from_quaternion(cls, q: Any) -> Rotation:
        """Create a rotation from a ``[w, x, y, z]`` quaternion."""
        return cls(q)

    @classmethod
    def from_rotation_matrix(cls, matrix: Any) -> Rotation:
        """Create a rotation from a 3x3 rotation matrix."""
        matrix_array = _as_finite_array(matrix, (3, 3), "rotation matrix")
        orth_err = np.max(np.abs(matrix_array @ matrix_array.T - np.eye(3)))
        require(
            bool(orth_err < 1e-6),
            f"rotation matrix must be orthogonal (max err={orth_err:.2e})",
        )
        det = np.linalg.det(matrix_array)
        require(bool(abs(det - 1.0) < 1e-6), "rotation matrix must have det=+1")
        rotation = SciPyRotation.from_matrix(matrix_array)
        return cls(_scipy_to_wxyz(rotation))

    @classmethod
    def from_euler(cls, a: float, b: float, c: float, convention: str) -> Rotation:
        """Create a rotation from Euler angles in radians."""
        _validate_convention(convention)
        rotation = SciPyRotation.from_euler(convention, [a, b, c])
        return cls(_scipy_to_wxyz(rotation))

    @classmethod
    def from_axis_angle(cls, axis: Any, angle: float) -> Rotation:
        """Create a rotation from a unit or non-zero axis and angle."""
        axis_array = _as_finite_array(axis, (3,), "axis")
        require_unit_vector(axis_array, "axis")
        rotation = SciPyRotation.from_rotvec(axis_array * float(angle))
        return cls(_scipy_to_wxyz(rotation))

    @classmethod
    def from_rodrigues(cls, vector: Any) -> Rotation:
        """Create a rotation from a Rodrigues/rotation vector."""
        vector_array = _as_finite_array(vector, (3,), "Rodrigues vector")
        rotation = SciPyRotation.from_rotvec(vector_array)
        return cls(_scipy_to_wxyz(rotation))

    def as_quaternion(self) -> np.ndarray:
        """Return a unit quaternion as ``[w, x, y, z]``."""
        return _scipy_to_wxyz(self._rotation)

    def as_rotation_matrix(self) -> np.ndarray:
        """Return a 3x3 rotation matrix."""
        return self._rotation.as_matrix()

    def as_euler(self, convention: str) -> tuple[float, float, float]:
        """Return Euler angles in radians for the given convention."""
        _validate_convention(convention)
        values = self._rotation.as_euler(convention)
        return (float(values[0]), float(values[1]), float(values[2]))

    def as_axis_angle(self) -> tuple[np.ndarray, float]:
        """Return ``(unit_axis, angle)``."""
        vector = self._rotation.as_rotvec()
        angle = float(np.linalg.norm(vector))
        if angle <= 1e-12:
            return np.array([1.0, 0.0, 0.0]), 0.0
        return vector / angle, angle

    def as_rodrigues(self) -> np.ndarray:
        """Return the Rodrigues/rotation vector."""
        return self._rotation.as_rotvec()

    def compose(self, other: Rotation) -> Rotation:
        """Compose this rotation with another: ``self * other``."""
        if not isinstance(other, Rotation):
            raise ValueError("other must be a Rotation")
        return Rotation(_scipy_to_wxyz(self._rotation * other._rotation))

    def inverse(self) -> Rotation:
        """Return the inverse rotation."""
        return Rotation(_scipy_to_wxyz(self._rotation.inv()))

    def __repr__(self) -> str:
        return f"Rotation(q={self.as_quaternion()})"

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Rotation):
            return NotImplemented
        q_self = self.as_quaternion()
        q_other = other.as_quaternion()
        return bool(
            np.allclose(q_self, q_other, atol=1e-10)
            or np.allclose(q_self, -q_other, atol=1e-10)
        )

    def __hash__(self) -> int:
        """Hash based on normalized quaternion (rounded to discrete bins).

        Two Rotation instances that compare equal must have the same hash.
        Since equality uses numerical tolerance (atol=1e-10), we round the
        quaternion to a coarse grid (1e-8 precision) for hashing.
        """
        q = self.as_quaternion()
        # Round to 8 decimal places for robust hashing across equivalent rotations
        q_rounded = tuple(np.round(q, 8).astype(float))
        return hash(q_rounded)


def _scipy_to_wxyz(rotation: SciPyRotation) -> np.ndarray:
    q_xyzw = rotation.as_quat()
    q_wxyz = np.array([q_xyzw[3], q_xyzw[0], q_xyzw[1], q_xyzw[2]], dtype=float)
    if q_wxyz[0] < 0:
        q_wxyz = -q_wxyz
    return q_wxyz


def _validate_convention(convention: str) -> None:
    if len(convention) != 3 or any(axis not in "xyzXYZ" for axis in convention):
        raise ValueError(f"Unknown Euler convention: {convention}")
